frontend/model/contracts/group.js

```diff
@@ -1267,6 +1280,47 @@ sbp('chelonia/defineContract', {
             groupContractID: contractID,
             chatRoomID: data.chatRoomID
           })
+
+          // The following queued invocation handles reference counting. It uses
+          // some transient (non-persistent) state to see whether the
+          // pairing `retain` to `release` was called. This situation could
+          // arise when syncing a group from scratch (e.g., from a new device)
+          // for chatrooms that have been joined and then left. In that case,
+          // there'll be no call to `retain` and therefore calling `release`
+          // is incorrect.
+          sbp('chelonia/queueInvocation', contractID, async () => {
+            // Has `retain` been skipped?
+            // Note: we check for 'skipped' instead of 'called' because
+            // side-effects cannot mutate state. The 'skipped' case can be
+            // checked without persistent storage, while the 'called' case
+            // would require some kind of persistent state (since the reference
+            // count is itself persisted)
+            const skippedRetain = sbp('okTurtles.data/get', `gi.contracts/group/chatroom-skipped-${contractID}-${data.chatRoomID}-${data.reference}`)
+            // There's no use of this transient value, so it's deleted to
+            // preserve memory
+            sbp('okTurtles.data/delete', `gi.contracts/group/chatroom-skipped-${contractID}-${data.chatRoomID}-${data.reference}`)
+
+            const { identityContractID } = sbp('state/vuex/state').loggedIn
+            // New user session
+            if (memberID !== identityContractID) {
+              return
+            }
+
+            const state = await sbp('chelonia/contract/state', contractID)
+
+            if (
+              state.chatRooms?.[data.chatRoomID]?.members[memberID]?.status === PROFILE_STATUS.ACTIVE ||
+              state.chatRooms?.[data.chatRoomID]?.members[memberID]?.joinedHeight !== data.reference ||
+              skippedRetain
+            ) {
+              return
+            }
+
+            // Release chatroom if we're sure `retain` was called
+            sbp('chelonia/contract/release', data.chatRoomID).catch(e => {
+              console.error(`[gi.contracts/group/leaveChatRoom/sideEffect] Error releasing chatroom ${data.chatRoomID}`, e)
+            })
+          })
```

Comment:
Did you say that it might be possible to simplify this code by adding a new feature to `okTurtles.eventQueue`? If so, let's do that.

Reply:
I don't have specific suggestions right now for the changes needed to `eventQueue`. I can work on a specific proposal.

Reply:
How much of an issue is this code? I originally thought that changing the `eventQueue` could work, but upon closer examination, I think that that alone can't fix it.

I've already described the situation in Slack, but the gist of this is that either:

  1. We need to increment / decrement the reference count without using any event queue, which would simplify the code but result in potentially unnecessary contract syncs (for example, you join and leave a group; the `retain` will result in syncing that group only for the `release` causing it to be removed, and the situation is worsened for multiple join / leaves, and also by chatrooms that are synced as a side effect of groups), or
  2. We keep on using an event queue to alleviate this concern. However, for the reference count to be correct, we need to know that `retain` was called before calling `release`. This needs to be somewhere in the state, and since side-effects can't change the contract state, I'm using `okTurtles.data`.

Reply:
If you have any questions about what the specific issue is, please let me know, because understanding it well is crucial to discussing a solution.

Reply:
I think an alternative solution is something I discussed earlier (which would, however, complicate reference counting).

The solution in this case would be that references, instead of being a number, which neeeds to be managed very carefully, are an identifier. This way, you have an explicit 'reason' for being subscribed, which can be removed. This is less error prone, but perhaps sounds more complicated (or not).

Needless to say, this would be a breaking change.